wormhole/file_transport.go

```diff
@@ -184,51 +184,88 @@ type fileTransport struct {
 	appID           string
 }
 
-func (t *fileTransport) connectViaRelay(myTransit *transitMsg, otherTransit *transitMsg) (net.Conn, error) {
+// combines hints and remove dublicates to minimaze connections
+func appendAndRemoveDuplicateHints(myTransit []transitHintsV1, peerTransit []transitHintsV1, hintType string) []transitHintsRelay {
+	var combinedHints []transitHintsRelay
+
+	for _, peerHints := range peerTransit {
+		if peerHints.Type == hintType {
+			combinedHints = peerHints.Hints
+
+		}
+
+	}
+
+	for _, myHints := range myTransit {
+		if myHints.Type == hintType {
+			for _, myHint := range myHints.Hints {
+				var dublicate bool = false
+				for _, peerHint := range combinedHints {
+					if myHint == peerHint {
+						dublicate = true
+					}
+				}
+				if !dublicate {
+					combinedHints = append(combinedHints, myHint)
+				}
+			}
+		}
+
+	}
+	return combinedHints
+}
+
+func (t *fileTransport) connectViaRelay(myTransit *transitMsg, peerTransit *transitMsg) (net.Conn, error) {
 
 	successChan := make(chan net.Conn)
 	failChan := make(chan string)
 
 	var count int
 
-	combinedHints := append(myTransit.HintsV1, otherTransit.HintsV1...)
-	for _, relay := range combinedHints {
-		if relay.Type == "relay-v1" {
-			for _, endpoint := range relay.Hints {
-				var relayUrl *url.URL
-				var err error
-
-				switch endpoint.Type {
-				case "direct-tcp-v1":
-					relayUrl = &url.URL{
-						Scheme: "tcp",
-						Host:   net.JoinHostPort(endpoint.Hostname, strconv.Itoa(endpoint.Port)),
-					}
-				case "websocket-v1":
-					relayUrl, err = url.Parse(endpoint.Url)
+	var cancelArray []context.CancelFunc
 
-				}
-				//ctx, cancel := context.WithCancel(context.Background())
-
-				//in case invalid url, cancel download
-				if err == nil {
-					ctx := context.Background()
-					count++
-					fmt.Printf("connectViaRelay: %s\n", relayUrl.String())
-					go t.connectToRelay(ctx, relayUrl, successChan, failChan)
-				} else {
-					//cancel()
-					continue
-				}
+	combinedHints := appendAndRemoveDuplicateHints(myTransit.HintsV1, peerTransit.HintsV1, "relay-v1")
+
+	for _, endpoint := range combinedHints {
+		var relayUrl *url.URL
+		var err error
+
+		switch endpoint.Type {
+		case "direct-tcp-v1":
+			relayUrl = &url.URL{
+				Scheme: "tcp",
+				Host:   net.JoinHostPort(endpoint.Hostname, strconv.Itoa(endpoint.Port)),
 			}
+		case "websocket-v1":
+			relayUrl, err = url.Parse(endpoint.Url)
+
+		}
+
+		if err == nil {
+			ctx, cancel := context.WithCancel(context.Background())
+
+			count++
+
+			go t.connectToRelay(ctx, relayUrl, successChan, failChan)
+			cancelArray = append(cancelArray, cancel)
+
+		} else {
+			continue
 		}
 	}
 
 	var conn net.Conn
+
 	for i := 0; i < count; i++ {
 		select {
 		case <-failChan:
 		case conn = <-successChan:
+			// cancel all other connections context
+			for j, cancel := range cancelArray {
+				if i != j {
+					cancel()
```

Comment:
One thing that came to my mind is that we call `cancel()` but are we catching that cancellation via `context.Done()` and doing something with it? I think cancelling context alone may not be good enough, we may have to explicitly close the connections.

Reply:
Not sure here. @ewanas @piegamesde can you take a look here too? 
Another way maybe, we could replace `ctx, cancel := context.WithCancel(context.Background())`
with `ctx, cancel := context.WithTimeout(ctx, 60*time.Seconds)`

Reply:
I'm a bit confused by the logic here in general; how does it make sure that you are cancelling at the correct index? My gut feeling is that this may be racy when multiple hints are involved.

Reply:
> I'm a bit confused by the logic here in general; how does it make sure that you are cancelling at the correct index? My gut feeling is that this may be racy when multiple hints are involved.

@piegamesde can you clarify this part.. and what would be your suggestion?

Reply:
If someone shuffles array, then yes, we can cancel successful connection. Thought to use map for cancelArray and relayUrl, but that is still unreliable: relayUrl may not match with value from successChan net.Conn.RemoteAddr(), for example to compare strings
```
TCP relay tcp://0.0.0.0:54743
RemoteAddr 127.0.0.1:54743
```

Reply:
My problem with this code is that I don't see any semantic connection between the counter `i` and the array index `j`. So I don't understand how this code manages to cancel the "right" connections.

Reply:
@vu3rdd @piegamesde I made some changes to be semantic correct, please review it carefully...

Reply:
> > t.directRecvHandshake(ctx, conn, successChan, failChan)
> 
> This function does not use the context at all.

Yes, this is not used here. We could remove probably. On connectToRelay it is required for DialContext function.  On connectViaRelay it is used only for cancellation of other go routines if one is successful.